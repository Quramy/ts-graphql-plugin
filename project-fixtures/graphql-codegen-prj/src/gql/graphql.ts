/* eslint-disable */
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
};

export type Post = {
  __typename?: 'Post';
  author: User;
  body: Scalars['String']['output'];
  createdAt: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  title: Scalars['String']['output'];
  updatedAt: Scalars['String']['output'];
};

export type Query = {
  __typename?: 'Query';
  popularPosts: Array<Post>;
};

export type User = {
  __typename?: 'User';
  avatarURL: Scalars['String']['output'];
  createdAt: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  name: Scalars['String']['output'];
  updatedAt: Scalars['String']['output'];
};

export type PopularPosts_QueryQueryVariables = Exact<{ [key: string]: never; }>;


export type PopularPosts_QueryQuery = { __typename?: 'Query', popularPosts: Array<(
    { __typename?: 'Post', id: string }
    & { ' $fragmentRefs'?: { 'PostSummary_PostFragment': PostSummary_PostFragment } }
  )> };

export type PostSummary_PostFragment = { __typename?: 'Post', id: string, title: string, author: (
    { __typename?: 'User', name: string }
    & { ' $fragmentRefs'?: { 'UserAvatar_UserFragment': UserAvatar_UserFragment } }
  ) } & { ' $fragmentName'?: 'PostSummary_PostFragment' };

export type UserAvatar_UserFragment = { __typename?: 'User', name: string, avatarURL: string } & { ' $fragmentName'?: 'UserAvatar_UserFragment' };

export const UserAvatar_UserFragmentDoc = {"kind":"Document","definitions":[{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}}]} as unknown as DocumentNode<UserAvatar_UserFragment, unknown>;
export const PostSummary_PostFragmentDoc = {"kind":"Document","definitions":[{"kind":"FragmentDefinition","name":{"kind":"Name","value":"PostSummary_Post"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"Post"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"author"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"UserAvatar_User"}}]}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}}]} as unknown as DocumentNode<PostSummary_PostFragment, unknown>;
export const PopularPosts_QueryDocument = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"PopularPosts_Query"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"popularPosts"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"PostSummary_Post"}}]}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"PostSummary_Post"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"Post"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"author"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"UserAvatar_User"}}]}}]}}]} as unknown as DocumentNode<PopularPosts_QueryQuery, PopularPosts_QueryQueryVariables>;