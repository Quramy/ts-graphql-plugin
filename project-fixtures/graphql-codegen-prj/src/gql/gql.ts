/* eslint-disable */
import * as types from './graphql';
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';

/**
 * Map of all GraphQL operations in the project.
 *
 * This map has several performance disadvantages:
 * 1. It is not tree-shakeable, so it will include all operations in the project.
 * 2. It is not minifiable, so the string of a GraphQL query will be multiple times inside the bundle.
 * 3. It does not support dead code elimination, so it will add unused operations.
 *
 * Therefore it is highly recommended to use the babel or swc plugin for production.
 */
const documents = {
    "\n  fragment Post_Post on Post {\n    title\n    author {\n      name\n    }\n  }\n": types.Post_PostFragmentDoc,
    "\n  query AppQuery {\n    popularPosts {\n      id\n      ...Post_Post\n    }\n  }\n": types.AppQueryDocument,
};

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 *
 *
 * @example
 * ```ts
 * const query = graphql(`query GetUser($id: ID!) { user(id: $id) { name } }`);
 * ```
 *
 * The query argument is unknown!
 * Please regenerate the types.
 */
export function graphql(source: string): unknown;

/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "\n  fragment Post_Post on Post {\n    title\n    author {\n      name\n    }\n  }\n"): (typeof documents)["\n  fragment Post_Post on Post {\n    title\n    author {\n      name\n    }\n  }\n"];
/**
 * The graphql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
 */
export function graphql(source: "\n  query AppQuery {\n    popularPosts {\n      id\n      ...Post_Post\n    }\n  }\n"): (typeof documents)["\n  query AppQuery {\n    popularPosts {\n      id\n      ...Post_Post\n    }\n  }\n"];

export function graphql(source: string) {
  return (documents as any)[source] ?? {};
}

export type DocumentType<TDocumentNode extends DocumentNode<any, any>> = TDocumentNode extends DocumentNode<  infer TType,  any>  ? TType  : never;