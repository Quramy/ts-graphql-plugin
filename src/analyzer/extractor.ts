import ts from 'typescript';
import { parse, print, DocumentNode, GraphQLError } from 'graphql';
import { visit } from 'graphql/language';
import { isTagged, ScriptSourceHelper, ResolvedTemplateInfo } from '../ts-ast-util';
import { ManifestOutput, ManifestDocumentEntry, OperationType } from './types';
import { ErrorWithLocation, ERROR_CODES } from '../errors';
import { detectDuplicatedFragments } from '../gql-ast-util';

export type ExtractorOptions = {
  removeDuplicatedFragments: boolean;
  scriptSourceHelper: ScriptSourceHelper;
  debug: (msg: string) => void;
};

export type ExtractTemplateResolveErrorResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  resolveTemplateError: {
    message: string;
    start: number;
    end: number;
  };
  documentNode: undefined;
  resolevedTemplateInfo: undefined;
  graphqlError: undefined;
};

export type ExtractGraphQLErrorResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  resolevedTemplateInfo: ResolvedTemplateInfo;
  resolveTemplateError: undefined;
  graphqlError: GraphQLError;
  resolveTemplateErrorMessage: undefined;
  documentNode: undefined;
};

export type ExtractSucceededResult = {
  fileName: string;
  templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression;
  documentNode: DocumentNode;
  resolevedTemplateInfo: ResolvedTemplateInfo;
  resolveTemplateError: undefined;
  graphqlError: undefined;
  resolveTemplateErrorMessage: undefined;
};

export type ExtractResult = ExtractTemplateResolveErrorResult | ExtractGraphQLErrorResult | ExtractSucceededResult;

export class Extractor {
  private readonly _removeDuplicatedFragments: boolean;
  private readonly _helper: ScriptSourceHelper;
  private readonly _debug: (msg: string) => void;

  constructor({ debug, removeDuplicatedFragments, scriptSourceHelper }: ExtractorOptions) {
    this._removeDuplicatedFragments = removeDuplicatedFragments;
    this._helper = scriptSourceHelper;
    this._debug = debug;
  }

  extract(files: string[], tagName?: string): ExtractResult[] {
    const results: ExtractResult[] = [];
    this._debug('Extract template literals from: ');
    this._debug(files.map(f => ' ' + f).join(',\n'));
    files.forEach(fileName => {
      const nodes = this._helper
        .getAllNodes(fileName, node => ts.isTemplateExpression(node) || ts.isNoSubstitutionTemplateLiteral(node))
        .filter(node => (tagName ? isTagged(node, tagName) : true)) as (
        | ts.TemplateExpression
        | ts.NoSubstitutionTemplateLiteral
      )[];
      nodes.forEach(node => {
        const { resolvedInfo, resolveErrors } = this._helper.resolveTemplateLiteral(fileName, node);
        if (!resolvedInfo) {
          resolveErrors
            .filter(re => re.fileName === fileName)
            .forEach(resolveError => {
              results.push({
                fileName,
                templateNode: node,
                resolveTemplateError: {
                  message: ERROR_CODES.templateIsTooComplex.message,
                  start: resolveError.start,
                  end: resolveError.end,
                },
                resolevedTemplateInfo: undefined,
                graphqlError: undefined,
                documentNode: undefined,
              });
            });
        } else {
          results.push({
            fileName,
            templateNode: node,
            resolevedTemplateInfo: resolvedInfo,
          } as any);
        }
      });
    });
    return results.map(result => {
      if (!result.resolevedTemplateInfo) return result;
      try {
        const rawDocumentNode = parse(result.resolevedTemplateInfo.combinedText);
        if (!this._removeDuplicatedFragments) {
          return {
            ...result,
            documentNode: rawDocumentNode,
          } as ExtractSucceededResult;
        }
        const duplicatedInfo = detectDuplicatedFragments(rawDocumentNode);
        const updatedResolvedInfo = duplicatedInfo.reduce(
          (acc, fragmentInfo) => this._helper.updateTemplateLiteralInfo(acc, fragmentInfo),
          result.resolevedTemplateInfo,
        );
        const documentNode = parse(updatedResolvedInfo.combinedText);
        return {
          ...result,
          documentNode,
          resolevedTemplateInfo: updatedResolvedInfo,
        } as ExtractSucceededResult;
      } catch (error) {
        if (error instanceof GraphQLError) {
          return {
            ...result,
            graphqlError: error,
          } as ExtractGraphQLErrorResult;
        } else {
          throw error;
        }
      }
    });
  }

  pickupErrors(
    extractResults: ExtractResult[],
    { ignoreGraphQLError }: { ignoreGraphQLError: boolean } = { ignoreGraphQLError: false },
  ) {
    const errors: ErrorWithLocation[] = [];
    extractResults.forEach(r => {
      if (r.resolveTemplateError) {
        errors.push(
          new ErrorWithLocation(r.resolveTemplateError.message, {
            fileName: r.fileName,
            severity: 'Warn',
            content: r.templateNode.getSourceFile().getFullText(),
            start: r.resolveTemplateError.start,
            end: r.resolveTemplateError.end,
          }),
        );
      } else if (!ignoreGraphQLError && r.graphqlError) {
        const innerLoc = r.graphqlError.locations && r.graphqlError.locations[0];
        if (!innerLoc) {
          errors.push(
            new ErrorWithLocation(r.graphqlError.message, {
              fileName: r.fileName,
              content: r.templateNode.getSourceFile().getFullText(),
              start: r.templateNode.getStart(),
              end: r.templateNode.getEnd(),
            }),
          );
        } else if (r.resolevedTemplateInfo) {
          const info = r.resolevedTemplateInfo;
          const innerPosition = info.convertInnerLocation2InnerPosition({
            line: innerLoc.line - 1,
            character: innerLoc.column - 1,
          });
          const start = info.getSourcePosition(innerPosition).pos;
          errors.push(
            new ErrorWithLocation(r.graphqlError.message, {
              fileName: r.fileName,
              content: r.templateNode.getSourceFile().getFullText(),
              start,
              end: start + 1,
            }),
          );
        }
      }
    });
    return errors;
  }

  getDominantDefinition(result: ExtractSucceededResult) {
    let type: OperationType | undefined;
    const definedFragmentNames: string[] = [];
    const referencedFragmentNames: string[] = [];
    let operationName: string | undefined;
    visit(result.documentNode, {
      FragmentDefinition(node) {
        if (!type) {
          type = 'fragment';
        }
        definedFragmentNames.push(node.name.value);
      },
      FragmentSpread(node) {
        referencedFragmentNames.push(node.name.value);
      },
      OperationDefinition(node) {
        if (!type || type === 'fragment') {
          type = node.operation;
        } else {
          type = 'complex';
        }
        if (!operationName) {
          operationName = node.name ? node.name.value : 'ANONYMOUS_QUERY';
        } else {
          operationName = 'MULTIPLE_OPERATIONS';
        }
      },
    });
    const noReferedFragmentNames = definedFragmentNames.filter(defName =>
      referencedFragmentNames.every(n => defName !== n),
    );
    return { type, operationName, fragmentName: noReferedFragmentNames[noReferedFragmentNames.length - 1] };
  }

  toManifest(extractResults: ExtractResult[], tagName: string = ''): ManifestOutput {
    const documents = extractResults
      .filter(r => !!r.documentNode)
      .map(result => {
        const r = result as ExtractSucceededResult;
        const { type, operationName, fragmentName } = this.getDominantDefinition(r);
        return {
          fileName: r.fileName,
          type: type || 'other',
          operationName,
          fragmentName,
          body: print(r.documentNode!),
          tag: tagName,
          templateLiteralNodeStart: this._helper.getLineAndChar(r.fileName, r.templateNode.getStart()),
          templateLiteralNodeEnd: this._helper.getLineAndChar(r.fileName, r.templateNode.getEnd()),
          documentStart: this._helper.getLineAndChar(r.fileName, r.templateNode.getStart() + 1),
          documentEnd: this._helper.getLineAndChar(r.fileName, r.templateNode.getEnd() - 1),
        } as ManifestDocumentEntry;
      });
    return {
      documents,
    };
  }
}
