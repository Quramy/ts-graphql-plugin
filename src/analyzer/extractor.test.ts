import { print } from 'graphql';
import { Extractor, ExtractSucceededResult } from './extractor';
import { createTesintExtractor } from './testing/testing-extractor';

describe(Extractor, () => {
  it('should extract GraphQL documents', () => {
    const extractor = createTesintExtractor([
      {
        fileName: 'main.ts',
        content: `
        import gql from 'graphql-tag';
        const query = gql\`
          query MyQuery {
            hello
          }
        \`;
        const mutation = gql\`
          mutation Greeting {
            greeting {
              reply
            }
          }
        \`;
      `,
      },
    ]);
    const result = extractor.extract(['main.ts'], 'gql');
    expect(result.map(r => print(r.documentNode!))).toMatchSnapshot();
  });

  it('should extract GraphQL documents and shrink duplicated fragments when removeDuplicatedFragments: true', () => {
    const extractor = createTesintExtractor(
      [
        {
          fileName: 'main.ts',
          content: `
        import gql from 'graphql-tag';
        const query = gql\`
          fragment A on Query {
            hello
          }
          fragment A on Query {
            hello
          }
          query MyQuery {
            ...A
          }
        \`;
      `,
        },
      ],
      true,
    );
    const result = extractor.extract(['main.ts'], 'gql');
    expect(result.map(r => print(r.documentNode!))).toMatchSnapshot();
  });

  it('should extract GraphQL documents and shrink duplicated fragments when removeDuplicatedFragments: false', () => {
    const extractor = createTesintExtractor(
      [
        {
          fileName: 'main.ts',
          content: `
        import gql from 'graphql-tag';
        const query = gql\`
          fragment A on Query {
            hello
          }
          fragment A on Query {
            hello
          }
          query MyQuery {
            ...A
          }
        \`;
      `,
        },
      ],
      false,
    );
    const result = extractor.extract(['main.ts'], 'gql');
    expect(result.map(r => print(r.documentNode!))).toMatchSnapshot();
  });

  it('should store template resolve errors with too complex interpolation', () => {
    const extractor = createTesintExtractor([
      {
        fileName: 'main.ts',
        content: `
          import gql from 'graphql-tag';
          const fragment = gql\`
            fragment MyFragment on Query {
              hello
            }
          \`;
          const query = gql\`
            \${fn(fragment)}
            query MyQuery {
              ...MyFragment
            }
          \`;
        `,
      },
    ]);
    const result = extractor.extract(['main.ts'], 'gql');
    expect(result[0].resolevedTemplateInfo).toBeTruthy();
    expect(result[1].resolevedTemplateInfo).toBeFalsy();
    expect(result[1].resolveTemplateError).toMatchSnapshot();
  });

  it('should store GraphQL syntax errors with invalid document', () => {
    const extractor = createTesintExtractor([
      {
        fileName: 'main.ts',
        content: `
        import gql from 'graphql-tag';
        const query = gql\`
          query MyQuery {
          }
        \`;
      `,
      },
    ]);
    const result = extractor.extract(['main.ts'], 'gql');
    expect(result[0].graphqlError).toBeTruthy();
  });

  it('should convert results to manifest JSON', () => {
    const extractor = createTesintExtractor([
      {
        fileName: 'main.ts',
        content: `
        import gql from 'graphql-tag';
        const query = gql\`
          query MyQuery {
            hello
          }
        \`;
        const mutation = gql\`
          mutation Greeting {
            greeting {
              reply
            }
          }
        \`;
      `,
      },
    ]);
    const result = extractor.extract(['main.ts'], 'gql');
    expect(extractor.toManifest(result)).toMatchSnapshot();
  });

  describe('getDominantDefinition', () => {
    it('should detect query type when document has only query', () => {
      const extractor = createTesintExtractor([
        {
          fileName: 'main.ts',
          content: `
            import gql from 'graphql-tag';
            const query = gql\`
              fragment X on Query {
                hello
              }
              query MyQuery {
                ...X
              }
            \`;
          `,
        },
      ]);
      const result = extractor.extract(['main.ts'], 'gql') as ExtractSucceededResult[];
      const { type, operationName } = extractor.getDominantDefinition(result[0]);
      expect(type).toBe('query');
      expect(operationName).toBe('MyQuery');
    });

    it('should detect mutation type when document has only mutation', () => {
      const extractor = createTesintExtractor([
        {
          fileName: 'main.ts',
          content: `
            import gql from 'graphql-tag';
            const query = gql\`
              mutation MyMutation {
                greeting {
                  reply
                }
              }
            \`;
          `,
        },
      ]);
      const result = extractor.extract(['main.ts'], 'gql') as ExtractSucceededResult[];
      const { type, operationName } = extractor.getDominantDefinition(result[0]);
      expect(type).toBe('mutation');
      expect(operationName).toBe('MyMutation');
    });

    it('should detect subscription type when document has only subscription', () => {
      const extractor = createTesintExtractor([
        {
          fileName: 'main.ts',
          content: `
            import gql from 'graphql-tag';
            const query = gql\`
              subscription MySubscription {
                greeting {
                  reply
                }
              }
            \`;
          `,
        },
      ]);
      const result = extractor.extract(['main.ts'], 'gql') as ExtractSucceededResult[];
      const { type, operationName } = extractor.getDominantDefinition(result[0]);
      expect(type).toBe('subscription');
      expect(operationName).toBe('MySubscription');
    });

    it('should return complex type with complex operations', () => {
      const extractor = createTesintExtractor([
        {
          fileName: 'main.ts',
          content: `
            import gql from 'graphql-tag';
            const query = gql\`
              subscription MySubscription {
                greeting {
                  reply
                }
              }
              mutation MyMutation {
                greeting {
                  reply
                }
              }
            \`;
          `,
        },
      ]);
      const result = extractor.extract(['main.ts'], 'gql') as ExtractSucceededResult[];
      const { type, operationName } = extractor.getDominantDefinition(result[0]);
      expect(type).toBe('complex');
      expect(operationName).toBe('MULTIPLE_OPERATIONS');
    });

    it('should detect fragment type when document has fragment and does not have any operations', () => {
      const extractor = createTesintExtractor([
        {
          fileName: 'main.ts',
          content: `
            import gql from 'graphql-tag';
            const query = gql\`
              fragment MyFragment on Query {
                hello
              }
            \`;
          `,
        },
      ]);
      const result = extractor.extract(['main.ts'], 'gql') as ExtractSucceededResult[];
      const { type, fragmentName } = extractor.getDominantDefinition(result[0]);
      expect(type).toBe('fragment');
      expect(fragmentName).toBe('MyFragment');
    });

    it('should detect fragment name to be exported', () => {
      const extractor = createTesintExtractor([
        {
          fileName: 'main.ts',
          content: `
            import gql from 'graphql-tag';
            const query = gql\`
              fragment MyFragment on Query {
                ...X
                hello
              }
              fragment X on Query {
                hello
              }
            \`;
          `,
        },
      ]);
      const result = extractor.extract(['main.ts'], 'gql') as ExtractSucceededResult[];
      const { type, fragmentName } = extractor.getDominantDefinition(result[0]);
      expect(type).toBe('fragment');
      expect(fragmentName).toBe('MyFragment');
    });

    it('should return the last fragments which are not referenced from others', () => {
      const extractor = createTesintExtractor([
        {
          fileName: 'main.ts',
          content: `
            import gql from 'graphql-tag';
            const query = gql\`
              fragment MyFragment1 on Query {
                ...X
                hello
              }
              fragment MyFragment2 on Query {
                ...X
                hello
              }
              fragment X on Query {
                hello
              }
            \`;
          `,
        },
      ]);
      const result = extractor.extract(['main.ts'], 'gql') as ExtractSucceededResult[];
      const { type, fragmentName } = extractor.getDominantDefinition(result[0]);
      expect(type).toBe('fragment');
      expect(fragmentName).toBe('MyFragment2');
    });
  });
});
