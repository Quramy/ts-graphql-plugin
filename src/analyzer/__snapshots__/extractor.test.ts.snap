// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extractor should convert results to manifest JSON 1`] = `
Object {
  "documents": Array [
    Object {
      "body": "query MyQuery {
  hello
}
",
      "documentEnd": Object {
        "character": 8,
        "line": 6,
      },
      "documentStart": Object {
        "character": 26,
        "line": 2,
      },
      "fileName": "main.ts",
      "fragmentName": undefined,
      "operationName": "MyQuery",
      "tag": "",
      "templateLiteralNodeEnd": Object {
        "character": 9,
        "line": 6,
      },
      "templateLiteralNodeStart": Object {
        "character": 25,
        "line": 2,
      },
      "type": "query",
    },
    Object {
      "body": "mutation Greeting {
  greeting {
    reply
  }
}
",
      "documentEnd": Object {
        "character": 8,
        "line": 13,
      },
      "documentStart": Object {
        "character": 29,
        "line": 7,
      },
      "fileName": "main.ts",
      "fragmentName": undefined,
      "operationName": "Greeting",
      "tag": "",
      "templateLiteralNodeEnd": Object {
        "character": 9,
        "line": 13,
      },
      "templateLiteralNodeStart": Object {
        "character": 28,
        "line": 7,
      },
      "type": "mutation",
    },
  ],
}
`;

exports[`Extractor should extract GraphQL documents 1`] = `
Array [
  "query MyQuery {
  hello
}
",
  "mutation Greeting {
  greeting {
    reply
  }
}
",
]
`;

exports[`Extractor should extract GraphQL documents and shrink duplicated fragments when removeDuplicatedFragments: false 1`] = `
Array [
  "fragment A on Query {
  hello
}

fragment A on Query {
  hello
}

query MyQuery {
  ...A
}
",
]
`;

exports[`Extractor should extract GraphQL documents and shrink duplicated fragments when removeDuplicatedFragments: true 1`] = `
Array [
  "fragment A on Query {
  hello
}

query MyQuery {
  ...A
}
",
]
`;

exports[`Extractor should store template resolve errors with too complex interpolation 1`] = `
Object {
  "end": 219,
  "message": "This operation or fragment has too complex interpolation to analyze.",
  "start": 207,
}
`;
